### ISSUE TITLE
[PATCH Compress-Zlib-1.08] Subroutine DynaLoader::dl_error    redefined on static build

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#1615]
Migrated from [rt.perl.org#2852](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=2852) (status was 'resolved')

Searchable as RT2852$

**Comment by p5pRT** [comment]
### From domo@computer.org

#### Created by domo@computer.org


\[Third shot\, in the hope that perlbug's up again\.\]

\(Yes\, this does look like another roughly contemporaneous report\.  But
it's different really\.\)

If perl is statically\-linked\, and in the absence of the patch below\,
make test for Compress&#8203;::Zlib gives

PERL\_DL\_NONLAZY=1 \./perl \-Iblib/arch \-Iblib/lib
\-I/usr/local/lib/perl5/5\.6\.0/pow
erpc\-machten \-I/usr/local/lib/perl5/5\.6\.0 \-e 'use Test&#8203;::Harness
qw\(&runtests $ve
rbose\); $verbose=0; runtests @&#8203;ARGV;' t/\*\.t
t/examples\.\.\.\.\.\.\.\.\.\.NOK 6Subroutine DynaLoader&#8203;::dl\_error redefined at
/usr/local
/lib/perl5/5\.6\.0/powerpc\-machten/DynaLoader\.pm line 93\.
t/examples\.\.\.\.\.\.\.\.\.\.NOK 13FAILED tests 2\, 4\, 6\, 8\, 11\, 13
&nbsp;         Failed 6/13 tests\, 53\.85% okay
t/zlib\.\.\.\.\.\.\.\.\.\.\.\.\.\.Subroutine DynaLoader&#8203;::dl\_error redefined at
/usr/local/lib/
perl5/5\.6\.0/powerpc\-machten/DynaLoader\.pm line 93\.
t/zlib\.\.\.\.\.\.\.\.\.\.\.\.\.\.ok
Failed Test  Status Wstat Total Fail  Failed  List of failed
\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
t/examples\.t                 13    6  46\.15%  2\, 4\, 6\, 8\, 11\, 13
Failed 1/2 test scripts\, 50\.00% okay\. 6/167 subtests failed\, 96\.41%
okay\.
make&#8203;: \*\*\* \[test\_static\] Error 255

Everything is fixed \(for me\) by stopping the module from requiring
DynaLoader\.  The patch does the same thing as
Change 5884 by gsar@&#8203;auger on 2000/03/22
did to the IO module\.  It needs testing on a dynamically\-linked
architecture\.

<details><summary>Inline Patch</summary>

```diff
--- Compress-Zlib-1.08/Zlib.pm~ Sun Jan  9 17:08:38 2000
+++ Compress-Zlib-1.08/Zlib.pm  Mon Mar 27 22:03:49 2000
@@ -12,7 +12,6 @@

  require 5.004 ;
  require Exporter;
-require DynaLoader;
  use AutoLoader;
  use Carp ;
  use IO::Handle ;
@@ -23,7 +22,7 @@

  $VERSION = "1.08" ;

-@ISA = qw(Exporter DynaLoader);
+@ISA = qw(Exporter);
  # Items to export into callers namespace by default. Note: do not

```
</details>
export
&nbsp;  \# names by default without a very good reason\. Use EXPORT\_OK instead\.
&nbsp;  \# Do not simply export all your public functions/methods/constants\.

<details><summary>Perl Info</summary>

```

Flags:
     category=library
     severity=medium

Site configuration information for perl v5.6.0:

Configured by domo at Thu Mar 23 20:29:52 WET 2000.

Summary of my perl5 (revision 5.0 version 6 subversion 0) configuration:
   Platform:
     osname=machten, osvers=4.1.4, archname=powerpc-machten
     uname='machten ppp100 5 0.5 powerpc '
     config_args='-des'
     hint=recommended, useposix=true, d_sigaction=define
     usethreads=undef use5005threads=undef useithreads=undef 
usemultiplicity=undef
     useperlio=undef d_sfio=undef uselargefiles=define
     use64bitint=undef use64bitall=undef uselongdouble=undef usesocks=undef
   Compiler:
     cc='cc', optimize='-O2 -fomit-frame-pointer', gccversion=2.8.1
     cppflags='-DNOTDEF_MACHTEN -DREG_INFTY=2047 -I/usr/local/include'
     ccflags ='-DNOTDEF_MACHTEN -DREG_INFTY=2047 -I/usr/local/include'
     stdchar='char', d_stdstdio=undef, usevfork=true
     intsize=4, longsize=4, ptrsize=4, doublesize=8
     d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=8
     ivtype='long', ivsize=4, nvtype='double', nvsize=8, 
Off_t='off_t', lseeksize=4
     alignbytes=8, usemymalloc=y, prototype=define
   Linker and Libraries:
     ld='ld', ldflags =' -Xlstack=1048576 -L/usr/local/lib'
     libpth=/usr/local/lib /usr/lib
     libs=-lndbm -lgdbm -ldb -lm -lc
     libc=/usr/lib/libc.a, so=so, useshrplib=false, libperl=libperl.a
   Dynamic Linking:
     dlsrc=dl_none.xs, dlext=none, d_dlsymun=undef, ccdlflags=''
     cccdlflags='', lddlflags=''

Locally applied patches:



@INC for perl v5.6.0:
     /usr/local/lib/perl5/5.6.0/powerpc-machten
     /usr/local/lib/perl5/5.6.0
     /usr/local/lib/perl5/site_perl/5.6.0/powerpc-machten
     /usr/local/lib/perl5/site_perl/5.6.0
     /usr/local/lib/perl5/site_perl
     .


Environment for perl v5.6.0:
     HOME=/home/domo
     LANG (unset)
     LANGUAGE (unset)
     LD_LIBRARY_PATH=/usr/lib
     LOGDIR (unset)
 
PATH=/sbin:/usr/sbin:/sbin:/usr/sbin:/home/domo/bin:/usr/local/bin:/usr/bin:/bin:/usr/X11/bin:/usr/libexec
     PERL_BADLANG (unset)
     SHELL=/bin/bash
-- 
Dominic Dunlop


```
</details>

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

Dominic Dunlop writes&#8203;:
> \(Yes\, this does look like another roughly contemporaneous report\.  But
> it's different really\.\)
> 
> If perl is statically\-linked\, and in the absence of the patch below\,
> make test for Compress&#8203;::Zlib gives

> t/examples\.\.\.\.\.\.\.\.\.\.NOK 6Subroutine DynaLoader&#8203;::dl\_error redefined at
> /usr/local
> /lib/perl5/5\.6\.0/powerpc\-machten/DynaLoader\.pm line 93\.

What if we Change DynaLoader to load dl\_error conditionally?

Ilya

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

At 13&#8203;:09 \-0500 2000\-03\-30\, Ilya Zakharevich wrote&#8203;:
>What if we Change DynaLoader to load dl\_error conditionally?

That would be nice&#8203;: I've found three modules which tickle this 
problem so far\.  No doubt there are more\.

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

Dominic Dunlop writes&#8203;:
> PERL\_DL\_NONLAZY=1 \./perl \-Iblib/arch \-Iblib/lib
> \-I/usr/local/lib/perl5/5\.6\.0/pow
> erpc\-machten \-I/usr/local/lib/perl5/5\.6\.0 \-e 'use Test&#8203;::Harness
> qw\(&runtests $ve
> rbose\); $verbose=0; runtests @&#8203;ARGV;' t/\*\.t
> t/examples\.\.\.\.\.\.\.\.\.\.NOK 6Subroutine DynaLoader&#8203;::dl\_error redefined at
> /usr/local

Now I'm puzzled why this happens\.  Here is how I did it / understand
&nbsp;  it&#8203;:

&nbsp;  DynaLoader&#8203;::dl\_error\(\) is defined in boot\_DynaLoader\(\)\.

&nbsp;  boot\_DynaLoader\(\) is \(almost completely\) autogenerated from dl\_\*\.xs\.

&nbsp;  boot\_DynaLoader\(\) can be called from XSLoader or DynaLoader\, but in
&nbsp;  both cases this call is done as

&nbsp;    package DynaLoader;
&nbsp;    boot\_DynaLoader\('DynaLoader'\) if defined\(&boot\_DynaLoader\) &&
&nbsp;                    \!defined\(&dl\_load\_file\);

&nbsp;  DynaLoader&#8203;::dl\_load\_file\(\) is defined simultaneously with
&nbsp;  DynaLoader&#8203;::dl\_error\(\)\.

Maybe in some dl\_\*\.xs dl\_load\_file\(\) is not defined\, but dl\_error\(\)
is?  Looks highly unprobable\.\.\.

Please inspect

&nbsp;  grep \-6 'dl\_error"' \*

in your ext/DynaLoader directory\.  I would expect to see
newXS\(whatever\) from DynaLoader\.c\, with both dl\_error and dl\_load\_file
loaded\.

Puzzled\,
Ilya

**Comment by p5pRT** [comment]
### From @doughera88

On Fri\, 31 Mar 2000\, Ilya Zakharevich wrote&#8203;:

> Dominic Dunlop writes&#8203;:

> > t/examples\.\.\.\.\.\.\.\.\.\.NOK 6Subroutine DynaLoader&#8203;::dl\_error redefined at


> Maybe in some dl\_\*\.xs dl\_load\_file\(\) is not defined\, but dl\_error\(\)
> is?  Looks highly unprobable\.\.\.

Excellent diagnosis\. That's exactly what happens\.  dl\_none\.xs does not
define dl\_load\_file\(\)\, but does define dl\_error\(\)\.

&nbsp;    Andy Dougherty        doughera@&#8203;lafayette\.edu

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

On Fri\, Mar 31\, 2000 at 03&#8203;:48&#8203;:20PM \-0500\, Andy Dougherty wrote&#8203;:
> > > t/examples\.\.\.\.\.\.\.\.\.\.NOK 6Subroutine DynaLoader&#8203;::dl\_error redefined at
> 
> 
> > Maybe in some dl\_\*\.xs dl\_load\_file\(\) is not defined\, but dl\_error\(\)
> > is?  Looks highly unprobable\.\.\.
> 
> Excellent diagnosis\. That's exactly what happens\.  dl\_none\.xs does not
> define dl\_load\_file\(\)\, but does define dl\_error\(\)\.

This patch is not tested\.

Ilya

<details><summary>Inline Patch</summary>

```diff
--- ./ext/DynaLoader/DynaLoader_pm.PL~	Sun Mar  5 23:56:02 2000
+++ ./ext/DynaLoader/DynaLoader_pm.PL	Fri Mar 31 15:50:44 2000
@@ -114,8 +114,9 @@ push(@dl_library_path, split(/:/, $ENV{L
 }
 
 # No prizes for guessing why we don't say 'bootstrap DynaLoader;' here.
+# dl_none.xs defines dl_error(), but no dl_load_file()
 boot_DynaLoader('DynaLoader') if defined(&boot_DynaLoader) &&
-                                !defined(&dl_load_file);
+                               !(defined(&dl_load_file) || defined(&dl_error));
 
 
 if ($dl_debug) {
--- ./ext/DynaLoader/XSLoader_pm.PL~	Wed Dec  8 01:23:10 1999
+++ ./ext/DynaLoader/XSLoader_pm.PL	Fri Mar 31 15:50:49 2000
@@ -39,8 +39,9 @@ print OUT <<'EOT';
 
 # No prizes for guessing why we don't say 'bootstrap DynaLoader;' here.
 package DynaLoader;
+# dl_none.xs defines dl_error(), but no dl_load_file()
 boot_DynaLoader('DynaLoader') if defined(&boot_DynaLoader) &&
-                                !defined(&dl_load_file);
+                               !(defined(&dl_load_file) || defined(&dl_error));
 package XSLoader;
 
 1; # End of main code
```
</details>

**Comment by p5pRT** [comment]
### From @gsar

On Thu\, 30 Mar 2000 09&#8203;:56&#8203;:58 \+0200\, Dominic Dunlop wrote&#8203;:
>If perl is statically\-linked\, and in the absence of the patch below\,
>make test for Compress&#8203;::Zlib gives
>
>PERL\_DL\_NONLAZY=1 \./perl \-Iblib/arch \-Iblib/lib
>\-I/usr/local/lib/perl5/5\.6\.0/pow
>erpc\-machten \-I/usr/local/lib/perl5/5\.6\.0 \-e 'use Test&#8203;::Harness
>qw\(&runtests $ve
>rbose\); $verbose=0; runtests @&#8203;ARGV;' t/\*\.t
>t/examples\.\.\.\.\.\.\.\.\.\.NOK 6Subroutine DynaLoader&#8203;::dl\_error redefined at
>/usr/local
>/lib/perl5/5\.6\.0/powerpc\-machten/DynaLoader\.pm line 93\.
>t/examples\.\.\.\.\.\.\.\.\.\.NOK 13FAILED tests 2\, 4\, 6\, 8\, 11\, 13
>         Failed 6/13 tests\, 53\.85% okay
>t/zlib\.\.\.\.\.\.\.\.\.\.\.\.\.\.Subroutine DynaLoader&#8203;::dl\_error redefined at
>/usr/local/lib/
>perl5/5\.6\.0/powerpc\-machten/DynaLoader\.pm line 93\.
>t/zlib\.\.\.\.\.\.\.\.\.\.\.\.\.\.ok
>Failed Test  Status Wstat Total Fail  Failed  List of failed
>\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
>t/examples\.t                 13    6  46\.15%  2\, 4\, 6\, 8\, 11\, 13
>Failed 1/2 test scripts\, 50\.00% okay\. 6/167 subtests failed\, 96\.41%
>okay\.
>make&#8203;: \*\*\* \[test\_static\] Error 255
>
>Everything is fixed \(for me\) by stopping the module from requiring
>DynaLoader\.  The patch does the same thing as
>Change 5884 by gsar@&#8203;auger on 2000/03/22
>did to the IO module\.  It needs testing on a dynamically\-linked
>architecture\.

Does this cure it?


Sarathy
gsar@&#8203;ActiveState\.com
<details><summary>Inline Patch</summary>

```diff
-----------------------------------8<-----------------------------------
Change 5947 by gsar@auger on 2000/04/25 18:11:53

	use &dl_error rather than &dl_load_file as the guard for calling
	boot_DynaLoader() (meant to fix dl_error() redefined warnings in
	statically built perl)

Affected files ...

... //depot/perl/ext/DynaLoader/DynaLoader_pm.PL#13 edit
... //depot/perl/ext/DynaLoader/XSLoader_pm.PL#2 edit

Differences ...

==== //depot/perl/ext/DynaLoader/DynaLoader_pm.PL#13 (text) ====
Index: perl/ext/DynaLoader/DynaLoader_pm.PL
--- perl/ext/DynaLoader/DynaLoader_pm.PL.~1~	Tue Apr 25 11:12:00 2000
+++ perl/ext/DynaLoader/DynaLoader_pm.PL	Tue Apr 25 11:12:00 2000
@@ -114,9 +114,9 @@
 }
 
 # No prizes for guessing why we don't say 'bootstrap DynaLoader;' here.
+# NOTE: All dl_*.xs (including dl_none.xs) define a dl_error() XSUB
 boot_DynaLoader('DynaLoader') if defined(&boot_DynaLoader) &&
-                                !defined(&dl_load_file);
-
+                                !defined(&dl_error);
 
 if ($dl_debug) {
     print STDERR "DynaLoader.pm loaded (@INC, @dl_library_path)\n";

==== //depot/perl/ext/DynaLoader/XSLoader_pm.PL#2 (text) ====
Index: perl/ext/DynaLoader/XSLoader_pm.PL
--- perl/ext/DynaLoader/XSLoader_pm.PL.~1~	Tue Apr 25 11:12:00 2000
+++ perl/ext/DynaLoader/XSLoader_pm.PL	Tue Apr 25 11:12:00 2000
@@ -37,10 +37,12 @@
 
 print OUT <<'EOT';
 
+package DynaLoader;
+
 # No prizes for guessing why we don't say 'bootstrap DynaLoader;' here.
-package DynaLoader;
+# NOTE: All dl_*.xs (including dl_none.xs) define a dl_error() XSUB
 boot_DynaLoader('DynaLoader') if defined(&boot_DynaLoader) &&
-                                !defined(&dl_load_file);
+                                !defined(&dl_error);
 package XSLoader;
 
 1; # End of main code
End of Patch.
```
</details>

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

At 11&#8203;:17 \-0700 2000\-04\-25\, Gurusamy Sarathy wrote&#8203;:
>Does this cure it?

Belatedly \(apologies\)\, yes\, that's fine\.  Thanks\.

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

This ticket can be closed&#8203;: patch 5947 and/or 7269 \(which appear to be 
the same patch on different branches\.  Or something\) fixed it\, along 
with two others\.
